### ISSUE TITLE
Multithreading question/request

### ISSUE THREAD

**Issue by BillZaumen** [oracle/graal#1717]
I'm posting this because of a possible migration issue regarding Nashorn.  I have a discrete event simulator that I've developed that can be configured using a scripting language. The
simulator provides 'factory' classes that can be configured in part by 'set' methods. In the simplest case, a set method takes two arguments: a name and a value, each represented by primitive types.  It can also takes three arguments: a name, a key, and a value, where the key is basically a key for a table with the specified name.  Using these methods is a bit tedious, so a facility exists to use a scripting language where (for example)  an ECMAScript object such as {"foo": 10, "bar": 20}, is turned into  factory.set("foo", 10); factory.set("bar", 20);

It is also possible to schedule events using threads, typically for tasks that run infrequently for
purposes such as instrumentation where one might collect some data and then 'sleep'  for some amount of simulation time, repeating this in a loop.  When run, however, only  a single thread runs at a time so locking and race conditions are not an issue. Starting threads and changing between threads are both handled in Java code. While one can code this example without using threads, or even an explicit loop, the code is not very clear: the code is more easily checked and debugged if a loop is used so coding it without the use of loops is justified only if there is a performance issue.

I really want this to work transparently.  I wrote the code initially when Rhino was the only choice, then modified it to work with both Rhino and Nashorn (the changes were mostly in how Java classes are imported and how to detect which ECMAScript implementation was being used).  It would be awkward (but worse, possibly error prone) if I cannot do everything using the standard Java scripting API.

So basically, the question is whether I can do this using Graalvm.  The documentation I found  seems to indicate that multiple threads are discouraged  without using multiple contexts, but it is not clear if the use of multiple contexts is enforced in some way when all the threading is handled by Java code that Graalvm would not have direct access to, and where only one thread is really active at any one time.  If Graalvm cannot handle this case, could Graalvm be changed to accommodate it?

**Comment by eleinadani** [comment]
Hi @BillZaumen,

Thanks for your question.

Two or more JavaScript contexts cannot be accessed concurrently (i.e., by two threads at the same time) in GraalVM JavaScript. Everything else is possible, though. For example, you can create a context, use it from one thread, and then use it from another thread later on. On the Java side, you just need proper synchronization to ensure that concurrent access to the same context does not happen.

[Here](https://medium.com/graalvm/multi-threaded-java-javascript-language-interoperability-in-graalvm-2f19c1f9c37b) you can find a blog post giving some more details about threading in GraalVM JS, while [here](https://github.com/graalvm/graaljs/tree/master/graal-js/src/com.oracle.truffle.js.test.threading/src/com/oracle/truffle/js/test/threading) we have some examples showing how JS contexts can be used in Java multi-threaded applications. Hopefully some of the examples already cover your scenario -- if not, we are happy to help.

BTW -- for JavaScript-specific issues, you can also open a ticket on the GraalVM JavaScript github repository: https://github.com/graalvm/graaljs

**Comment by BillZaumen** [comment]
On Thu, 2019-10-03 at 09:33 -0700, Daniele Bonetta wrote:
> Hi @BillZaumen,
> Thanks for your question.
> Two or more JavaScript contexts cannot be accessed concurrently
> (i.e., by two threads at the same time) in GraalVM JavaScript.
> Everything else is possible, though. For example, you can create a
> context, use it from one thread, and then use it from another thread
> later on. On the Java side, you just need proper synchronization to
> ensure that concurrent access to the same context does not happen.
> Here you can find a blog post giving some more details about
> threading in GraalVM JS, while here we have some examples showing how
> JS contexts can be used in Java multi-threaded applications.
> Hopefully some of the examples already cover your scenario -- if not,
> we are happy to help.
> BTW -- for JavaScript-specific issues, you can also open a ticket on
> the GraalVM JavaScript github repository: 
> https://github.com/graalvm/graaljs
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Thanks for replying!  In case it is useful as an example, I've included
some code that shows what my simulator is doing (this is a
simplification  - the simulator handles multiple types of events and
events can be canceled or rescheduled).  The example shows event-queue
processing being started from a script - this is a common use case -
and uses simplified data structures to keep the example short. So far,
I've just tried it with Nashorn (the default for the version of openjdk
I'm running).  The idea is that each TaskThread has two monitors, one
for  controlling the scheduler and one for controlling its thread (in
the simulator, some events don't require threads at all, and since this
is a common case, I didn't want any monitors in the scheduler itself).
The result is that only one thread executes ECMAScript statements at a
time. Regardless of where the scheduler starts, it will start a
separate thread or notify that thread so it can continue, and then wait
for a notification that the thread has either run to completion or
paused.  Upon receiving that notification, the scheduler will go on to
the next event. 

The Java code is the following:
------------- test/ThreadTest.java ------------
package test;

import java.io.Reader;
import java.io.FileReader;
import java.nio.charset.Charset;
import java.util.*;
import javax.script.*;
   
public class ThreadTest {

    static ScriptEngine engine;

    public static class TaskThread extends Thread {
	
	Queue<Runnable> scheduler;

	Runnable runnable;
	boolean runnablePaused = true;
	Object runnableMonitor = new Object();

	boolean schedPaused = false;
	Object schedMonitor = new Object();

	public TaskThread(Queue<Runnable> scheduler, final Object obj)
{
	    super();
	    this.scheduler = scheduler;
	    runnable = () -> {
		try {
		    ((Invocable)engine).invokeMethod(obj, "run");
		} catch (Exception se) {
		    se.printStackTrace();
		}
	    };
	}

	public void start() {
	    schedPaused = true;
	    runnablePaused = true;
	    super.start();
	}

	public void run() {
	    try {
		runnable.run();
	    } finally {
		synchronized(this.schedMonitor) {
		    this.schedPaused = false;
		    this.schedMonitor.notifyAll();
		}
	    }
	}

	public static void pause() {
	    final TaskThread taskThread =
		(TaskThread) (Thread.currentThread());
	    taskThread.scheduler.add(() -> {
		    try {
			synchronized(taskThread.runnableMonitor) {
			    taskThread.runnablePaused = false;
			    taskThread.runnableMonitor.notifyAll();
			}
		    } catch (Exception e) {
			e.printStackTrace();
		    }
		});
	    try {
		synchronized(taskThread.schedMonitor) {
		    taskThread.schedPaused = false;
		    taskThread.schedMonitor.notifyAll();
		}
		synchronized (taskThread.runnableMonitor) {
		    while (taskThread.runnablePaused) {
			taskThread.runnableMonitor.wait(); 
		    }
		}
	    } catch (Exception e) {
		e.printStackTrace();
	    }
	}


    }

    static Queue<Runnable> sched = new LinkedList<Runnable>();

    public static void addThread(TaskThread thread) {
	sched.add(() -> {
		thread.start();
		try {
		    synchronized(thread.schedMonitor) {
			while (thread.schedPaused) {
			    thread.schedMonitor.wait();
			}
		    }
		} catch (Exception e) {
		    e.printStackTrace();
		}
	    });
    }

    public static TaskThread newThread(Object obj) {
	return new TaskThread(sched, obj);
    }

    public static void run() {
	Runnable r;
	while ((r = sched.poll()) != null) {
	    r.run();
	}
    }

    public static void main(String argv[]) throws Exception {
	ScriptEngineManager manager = new ScriptEngineManager();
	engine = manager.getEngineByName("ECMAScript");
	Reader reader = new FileReader("test.js",
				       Charset.forName("UTF-8"));
	engine.eval(reader);
    }
}
------------ test.js -------------
var ThreadTest = Java.type("test.ThreadTest");
var TaskThread = Java.type("test.ThreadTest.TaskThread");

function newRunnable(f) {
    var result = {
	run: f
    };
    return result;
}

var object1 = newRunnable(function() {
    print("hello");
    TaskThread.pause();
    print("goodbye");
});

var object2 = newRunnable(function() {
    print("Hello");
    TaskThread.pause();
    print("Goodbye");
});

ThreadTest.addThread(ThreadTest.newThread(object1));
ThreadTest.addThread(ThreadTest.newThread(object2));

ThreadTest.run();
------------------------------------------------------------

**Comment by wirthi** [comment]
Closing this stale ticket. If still relevant, please open a new ticket for a recent GraalVM version on https://www.github.com/oracle/graaljs, thanks!

Documentation on Multithreading in Graal.js can be found in https://github.com/oracle/graaljs/blob/master/docs/user/Multithreading.md
